nixos: replace activationScripts 2/x

nikstur opened the pull request:
## Description of changes

Replace some more relatively easy to replace activationScripts. This is the second PR of the series started here: https://github.com/NixOS/nixpkgs/pull/263203 

It's part of the broader "Perlless Activation" endeavour: https://github.com/NixOS/nixpkgs/issues/267982

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nikstur committed:
nixos/vmware-host: replace activationScript

Replace with a dedicated system servie ordered before the other VMWare
services.

nikstur committed:
nixos/logcheck: replace activationScript via tmpfiles

nikstur committed:
nixos/amdgpu-pro: replacce activationScript via tmpfiles

nikstur committed:
nixos/tests/incron: replace activcationScript via tmpfiles

nikstur committed:
nixos/trackpoint: replace activationScript

Replace with a separate systemd service ordered before sysinit.target

nikstur committed:
nixos/yggdrasil: replace activationScript

Replace with separate service because it cannot be moved into the
preStart of the yggdrasil service.

nikstur committed:
nixos/tests/nextcloud: replace activationScript

Replace with a seprate systemd service

nikstur committed:
nixos/wrappers: order service after sysusers service

nikstur committed:
nixos/binfmt: remove unused stringAfter

nikstur committed:
nixos/nix-channel: remove unused stringAfter

nikstur committed:
nixos/borgbackup: replace activationScript via tmpfiles

nikstur committed:
nixos/ipa: replace activationScript

Replaced with a dedicated systemd service.

nikstur committed:
nixos/ldap: replace activationScript

Replace with a dedicated systemd service.

philiptaron reviewed (approved):
Great.

philiptaron commented on nixos/modules/config/ldap.nix:
```diff
@@ -258,42 +246,60 @@ in
+        ldap-password = {
+          wantedBy = [ "sysinit.target" ];
+          before = [ "sysinit.target" ];
+          unitConfig.DefaultDependencies = false;
```
I'm also curious why this isn't a `serviceConfig.Type = "oneshot";`. It looks one-shotty to me.

nikstur replied:
Anything that is ordered before sysinit.target needs to have default dependencies disabled. Otherwise it just doesnt run: https://www.freedesktop.org/software/systemd/man/latest/systemd.special.html#sysinit.target

Yes I think you're right this is a oneshot service.

philiptaron commented on nixos/modules/security/ipa.nix:
```diff
@@ -181,25 +181,30 @@ in {
+          cp ${cfg.certificate} /etc/ipa/ca.crt
+        fi
+
+        if [ ! -f /etc/krb5.keytab ]; then
+          cat <<EOF
+
+            In order to complete FreeIPA integration, please join the domain by completing the following steps:
+            1. Authenticate as an IPA user authorized to join new hosts, e.g. kinit admin@${cfg.realm}
+            2. Join the domain and obtain the keytab file: ipa-join
+            3. Install the keytab file: sudo install -m 600 krb5.keytab /etc/
+            4. Restart sssd systemd service: sudo systemctl restart sssd
+
+        EOF
+        fi
```
Future: this looks like a `tmpfiles` `f` rule.

philiptaron commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -142,16 +142,23 @@ in
+      systemd.services.yggdrasil-persistent-keys = lib.mkIf cfg.persistentKeys {
+        wantedBy = [ "multi-user.target" ];
+        before = [ "yggdrasil.service" ];
+        unitConfig.DefaultDependencies = false;
```
Yadda on the `oneshot` and `DefaultDependencies=false` questions.

I'll omit these comments for the rest of the review.

nikstur replied:
Here `DefaultDependencies=false` actually wasn't needed. I removed it.

philiptaron commented on nixos/tests/nextcloud/basic.nix:
```diff
@@ -13,10 +13,15 @@ in {
-        echo "http://nextcloud/remote.php/dav/files/${adminuser} ${adminuser} ${adminpass}" > /tmp/davfs2-secrets
-        chmod 600 /tmp/davfs2-secrets
-      '';
+      systemd.services.davfs2-secrets = {
+        wantedBy = [ "sysinit.target" ];
+        before = [ "sysinit.target" ];
+        unitConfig.DefaultDependencies = false;
+        script = ''
+          echo "http://nextcloud/remote.php/dav/files/${adminuser} ${adminuser} ${adminpass}" > /tmp/davfs2-secrets
+          chmod 600 /tmp/davfs2-secrets
+        '';
+      };
```
I think this is `systemd.tmpfiles` `f`

philiptaron reviewed (approved):
Some curiosities, but this looks good to me.

nikstur commented on nixos/modules/config/ldap.nix:
```diff
@@ -258,42 +246,60 @@ in
+        nslcd = {
+          wantedBy = [ "multi-user.target" ];
+
+          preStart = ''
+            umask 0077
+            conf="$(mktemp)"
+            {
+              cat ${nslcdConfig}
+              test -z '${cfg.bind.distinguishedName}' -o ! -f '${cfg.bind.passwordFile}' ||
+              printf 'bindpw %s\n' "$(cat '${cfg.bind.passwordFile}')"
+              test -z '${cfg.daemon.rootpwmoddn}' -o ! -f '${cfg.daemon.rootpwmodpwFile}' ||
+              printf 'rootpwmodpw %s\n' "$(cat '${cfg.daemon.rootpwmodpwFile}')"
+            } >"$conf"
+            mv -fT "$conf" /run/nslcd/nslcd.conf
+          '';
```
None of these services have tests, so I'm more hestitant to change them to tmpfiles rules. However, if you're interesting in doing that I'd be glad to help review!

nikstur commented on nixos/modules/tasks/trackpoint.nix:
```diff
@@ -80,10 +80,14 @@ with lib;
-      system.activationScripts.trackpoint =
-        ''
-          ${config.systemd.package}/bin/udevadm trigger --attr-match=name="${cfg.device}"
+      systemd.services.trackpoint = {
+        wantedBy = [ "sysinit.target" ] ;
+        before = [ "sysinit.target" ] ;
+        unitConfig.DefaultDependencies = false;
+        serviceConfig.ExecStart = ''
+          ${config.systemd.package}/bin/udevadm trigger --attr-match=name="${cfg.device}
         '';
+      };
```
If you have a way of verifying this, I'd be glad to review!

nikstur commented on nixos/modules/services/networking/yggdrasil.nix:
```diff
@@ -142,16 +142,23 @@ in
+          if [ ! -e ${keysPath} ]
+          then
+            mkdir --mode=700 -p ${builtins.dirOf keysPath}
+            ${binYggdrasil} -genconf -json \
+              | ${pkgs.jq}/bin/jq \
+                  'to_entries|map(select(.key|endswith("Key")))|from_entries' \
+              > ${keysPath}
```
Probably

SuperSandro2000 commented on nixos/modules/hardware/video/amdgpu-pro.nix:
```diff
@@ -39,9 +39,14 @@ in
-      ln -sfn ${package}/opt/amdgpu{,-pro} /run
-    '';
+    systemd.tmpfiles.settings.amdgpu-pro = {
+      "/run/amdgpu"."L+" =  {
+        argument = "${package}/opt/amdgpu";
+      };
+      "/run/amdgpu-pro"."L+" =  {
+        argument = "${package}/opt/amdgpu-pro";
+      };
```
```suggestion
      "/run/amdgpu"."L+".argument = "${package}/opt/amdgpu";
      "/run/amdgpu-pro"."L+".argument = "${package}/opt/amdgpu-pro";
```
I am a bit confused why the normal tmpfiles.d syntax is not used 🤔

lheckemann replied:
This is a new way to specify tmpfiles rules, which allows overriding thanks to the use of attrsets instead of lists.

SuperSandro2000 commented on nixos/tests/nextcloud/basic.nix:
```diff
@@ -13,10 +13,17 @@ in {
+        serviceConfig.Type = "oneshot";
+        serviceConfig.RemainAfterExit = true;
+        script = ''
+          echo "http://nextcloud/remote.php/dav/files/${adminuser} ${adminuser} ${adminpass}" > /tmp/davfs2-secrets
+          chmod 600 /tmp/davfs2-secrets
```
out of scope for this PR but unless the service is using private tmp then this is not that secure

SuperSandro2000 reviewed (commented):
would have love to give some deeper feedback but the changes are a bit tricky for me :/

nikstur commented:
@ofborg test yggdrasil nextcloud.basic26 wrappers borgbackup incron

nikstur commented:
I added shutdown.target dependencies to the sysinit services to be aligned with https://github.com/NixOS/nixpkgs/pull/271326 and converted the nextcloud service to tmpfiles.

philiptaron reviewed (commented):
So glad to see this land.
